src/result.ts

```diff
@@ -115,19 +365,21 @@ export function toResult<T, E>(fn: () => T): Result<T, E> {
     }
 }
 
-export async function toResultAsync<T, E>(tg: Promise<T> | (() => Promise<T>)): Promise<Result<T, E>> {
-    if (typeof tg !== 'function' && typeof tg !== 'object') {
-        throw new Error('toResultAsync accepts only promises or functions that return a promise');
+type Fn<T> = (...params: any[]) => T;
+type ResFn<T, E, F extends Fn<T>> = (...params: Parameters<F>) => Result<T, E>;
+/**
+ * @description Catches exceptions and converts them into Result
+*/
+export function resultify<T, E, F extends Fn<T>>(fn: F, mapErr: (err: unknown) => E): ResFn<T, E, F> {
+    return (...params) => { 
+        try {
+            const val = fn(...params);
+            return Ok(val);
+        } catch(e) {
+            return Err(mapErr(e));
+        }
     }
-
```

Comment:
I have fixed the first and second points. I would like to discuss the third one with you.

The error passed to mapErr directly translates to the error caught by resultify. That error can safely only be typed unknown, or unsafely any.

See this example
``` ts
const test = () => {
    try {
        throw 1;
    } catch (err: number) { // Catch clause variable type annotation must be 'any' or 'unknown' if specified. [TS1196]
        return err.message;
    }
}
```
If the error would be any, user could specify type, but it would be unsafe. Like this user has to narrow his error type to ensure typesafe E of his Results

Reply:
I have thought about that and came to a solution that satisfies both the convenience and safety. We can reduce all error types to one.

The result code will look like this:
```ts
export function resultify<F extends Fn<ReturnType<F>>>(fn: F, mapErr = (e: BaseError) => e) {
    return (...params: any[]) => { 
        try {
            const val = fn(...params);
            return Ok(val);
        } catch(e) {
            const baseErr = new BaseError(e, '');
            return Err(mapErr(baseErr));
        }
    }
}

/**
 * @description converts non Error values to Error
 */
export class BaseError extends Error {
    /**
     * @description exotic value that was thrown
     */
    public origValue: unknown;

    name = "BaseError";

    // specified any to not write additional casts
    constructor(origValue: any, msg: string) {
        super(msg);

        const isErr = origValue instanceof Error;

        if (!isErr) {
            this.origValue = origValue;
        }

        // setting origVal.toString() into message
        if (typeof origValue?.toString == 'function' && !isErr) {
            const newMsg = origValue.toString();

            if (typeof newMsg === 'string') {
                const type = Array.isArray(origValue) ? 'array' : typeof origValue;
                this.message = `Caught exotic value (${type}): ${newMsg}`;
            }
        }

        // copy orig error params
        if (isErr) {
            this.stack = origValue.stack;
            this.message = origValue.message;
            this.name = origValue.name;
        }
    }
}
```

Reply:
I really like the idea, but there is one issue why we shouldn't do it exactly like this. For example `exec` from `child_process` rejects with `ExecException` which is instance of `Error` but also has properties `killed`, `cmd`, `code`, etc... These additional properties would be lost here.

So I thought maybe we could just passthrough thrown Errors, and wrap only non-Error thrown values. But then accessing the original thrown value would be quite awkward. For example
```
if ('origValue' in err) {
    console.log(err.origValue);
}
```

But maybe that is not really an issue...? Because honestly, how often does someone throw non-Error anyway, right...? So that is what I went with for now.

Reply:
You can easily throw a string, that's a pretty common case, so we can't ignore it. And yep, we still can provide a better solution. We will check & change prototype if needed, also copy the original properties.  

Js is one of the most powerful languages. The code:

```ts
import clone from "clone";

/**
 * @description converts non Error values to Error
 */
export class BaseError extends Error {
    /**
     * @description exotic value that was thrown
     */
    public origValue?: unknown;

    name = "BaseError";

    // specified any to not write additional casts
    constructor(errVal: any, msg: string) {
        super(msg);

        const isErr = errVal instanceof Error;

        if (!isErr) {
            this.origValue = errVal;
        }

        // setting origVal.toString() into message
        if (typeof errVal?.toString == 'function' && !isErr) {
            const newMsg = errVal.toString();

            if (typeof newMsg === 'string') {
                const type = Array.isArray(errVal) ? 'array' : typeof errVal;
                this.message = `Caught exotic value (${type}): ${newMsg}`;
            }
        }

        // copy orig error params
        if (isErr) {
            const errProto = Object.getPrototypeOf(errVal);

            // Copy prototype and properties if errVal is
            // a subclass of Error
            if (errProto !== Error.prototype) {
                Object.setPrototypeOf(this, errProto);

                const copy = clone(errVal);
                Object.assign(this, copy);
            }

            this.stack = errVal.stack;
            this.message = errVal.message;
            this.name = errVal.name;
        }
    }
}
```

Reply:
I think this is not a good solution.

1. We are forcing everyone to build stack trace, any time they throw non-Error. Building stack trace can be very expensive.
2. We are also building second stack trace when Error is thrown only to discard it.
3. To know whether "exotic value" was thrown, user still has to check for `'origValue' in err`, so no improvement over current code. User could now also check if `origValue` is undefined, but that is less correct, since `undefined` could be thrown and assigned to `origValue`.
5. User has no way to know that their error typed `BaseError` can be `instanceof` any other `Error` based class. Unlike `unknown`, `BaseError` does not feel like something you can narrow down to another type.
6. People will expect `BaseError` in their err, and suddenly see in their log `ExecException`. That seems to be very high surprise factor.
7. I do not think there is any real benefit over current code.

I think that err in case of (async)resultify and fromPromise should just remain unknown. We should let user handle their errors, and not do anything obscure and unexpected.  I would accept wrapping non-Error values because I think that throwing non-Errors is very niche, but even that feels wrong because of point 1. If you are worried about convenience, know that both resultify and fromPromise are intended only to aid in starting using of Results in existing codebases without having to rewrite existing functions for which one would want to use Results. New functions should already be returning Result/AsyncResult where both T and E are already correctly typed. Also writing custom helper to wrap error in whatever container is trivial and in my opinion should not be part of this library as it takes control/freedom away from the user of this library.

TLDR; I would prefer reverting `err` back to `unknown`, but would also accept only wrapping non-Errors.
Please, let me know about your opinion on the subject.

Reply:
The code above allowed type narrowing with 'instanceof', but, never mind, the inheritance from Error introduces too many issues with the stack trace. You are absolutely right about just passing the original error instance; there is no point in modifying it.

I think we have a better way to handle exotic error values. Simply create a plain object that implements the Error interface instead of creating an Error instance. This approach eliminates the computation of a stack trace pointing to the library code and improves performance.

I don't see any serious problems in this simpler implementation.
```ts
type BaseError = Error & {
    /**
     * @description exotic value that was thrown
     */
    origValue?: unknown;
}

const toBaseError = (errVal: any): BaseError => {
    const isErr = errVal instanceof Error;

    if (isErr) {
        // regular error, just pass it throw
        return errVal as BaseError;
    }

    // Error value is not an error instance
    const resErr = {} as BaseError;

    resErr.origValue = errVal;
    resErr.name = 'BaseError';

    // setting origVal.toString() into message
    if (typeof errVal?.toString == 'function') {
        const newMsg = errVal.toString();

        if (typeof newMsg === 'string') {
            const type = Array.isArray(errVal) ? 'array' : typeof errVal;
            resErr.message = `Caught a plain value (${type}): ${newMsg}`;
        }
    }

    return resErr;
};
```

Reply:
I am happy to hear that we are on the same page now. My implementation differs very slightly from yours, but functionally it should be the same.  I think this should resolve the last issue of this PR?

Reply:
All great, only one detail left. 

The `Type ResultBaseError = BaseError | Error;` will not show the existence of the 'origValue' property. The only way for a user to get know about it - go to the function definition.

This will fix the problem:
```ts
export type ResultBaseError = Error & {
    origValue?: unknown;
};
```